experimental/dds/tree/src/interface-prototype/Checkout.ts

```diff
@@ -0,0 +1,276 @@
+isTrue<isAssignableTo<Range, RangeViewReadonly>>();
+isTrue<isAssignableTo<TreeNode, TreeNodeViewReadonly>>();
+
+isTrue<areSafelyAssignable<Place, Contextualize<Place>>>();
+isTrue<areSafelyAssignable<Range, Contextualize<Range>>>();
+isTrue<areSafelyAssignable<TreeNode, Contextualize<TreeNode>>>();
+
+isTrue<areSafelyAssignable<Place, Contextualize<PlaceData>>>();
+isTrue<areSafelyAssignable<Range, Contextualize<RangeData>>>();
+isTrue<areSafelyAssignable<TreeNode, Contextualize<TreeNodeData>>>();
+
+isTrue<areSafelyAssignable<PlaceViewReadonly, ContextualizeView<Place>>>();
+isTrue<areSafelyAssignable<RangeViewReadonly, ContextualizeView<Range>>>();
+isTrue<areSafelyAssignable<TreeNodeViewReadonly, ContextualizeView<TreeNode>>>();
+
+isTrue<areSafelyAssignable<PlaceViewReadonly, ContextualizeView<PlaceData>>>();
+isTrue<areSafelyAssignable<RangeViewReadonly, ContextualizeView<RangeData>>>();
+isTrue<areSafelyAssignable<TreeNodeViewReadonly, ContextualizeView<TreeNodeData>>>();
+
+type ContextualizeView<TData extends AnchorData> = TData extends PlaceData
+	? PlaceViewReadonly
+	: TData extends RangeData
+	? RangeViewReadonly
+	: TData extends TreeNodeData
+	? TreeNodeViewReadonly
+	: Anchor;
+
+type Decontextualize<TData extends AnchorData> = TData extends PlaceData
+	? PlaceData
+	: TData extends RangeData
+	? RangeData
+	: TData extends TreeNodeData
+	? TreeNodeData
+	: AnchorData;
+
+//////////////////////////////////////////////////////////////////////////////////
+
+// Document and Snapshots
+
+export interface SharedTree extends Tree {
+	// createBranch(...);
+	// mergeBranch(...);
+	// deleteBranch(...);
+	// squash(firstRevision: ShortId, lastRevision: ShortId); // Revision id is changeId of preceding edit
+	// addSnapshot(revision: ShortId): SharedTreeSnapshot;
+	// getSnapshot(revision: ShortId): SharedTreeSnapshot;
+	// removeSnapshot(revision: ShortId): void;
+	// readonly firstRevision: ShortId;  // Or a constant value, e.g. 0x1?
+	// getChange(changeId: ShortId): TreeNode;
+	// nextChange(changeId: ShortId): ShortId; // Must take branching into account once we have that
+	// previousChange(changeId: ShortId): ShortId;
+	// createDiff(...);
+	// ...
+
+	// Synchronously produce a checkout where placeholders show up wherever they are in the actual data.
+	checkout(r: CommandRegistry): Checkout;
+	// Get a checkout with control over where the placeholders are.
+	checkout(r: CommandRegistry, prefetchFilter: PrefetchFilter): Promise<Checkout>;
+	// TODO: maybe option to get checkout with different API for when enforcing no placeholders at all (aka prefetching everything)
+}
+
+/**
+ * Specifies what data to prefetch.
+ * Where the filters return false, Placeholders are allowed.
+ * Returning true for all with fetch everything and produce no placeholders.
+ */
+export interface PrefetchFilter {
+	value: Filter<Definition>;
+	children: Filter<Definition>;
+	traitChunks: Filter<TraitLabel>;
+}
```

Comment:
I think you'll want something declarative that you can ship to the server like, [JMESPath](https://jmespath.org/).

Reply:
Fluid blob storage does not support such things. We are relying on somewhat large logical tree chunks to keep the round trip count low.

If we do end up deploying custom server code to reduce round trip delays when prefetching, something like that could be a good idea.

That said, I think this is yet another augment for having a schema based typed API which included chunking.
Schemas which include prefetch/async requirements, are a declarative query language which work well here.